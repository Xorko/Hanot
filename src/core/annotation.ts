interface Annotation {
  readonly value: string;
}

export type Type = Annotation;
