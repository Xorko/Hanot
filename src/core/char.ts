/*
 * A Char represents a content of an annotation unit, it could be some noise, a letter or something of pending char
 */

interface Noise {
  readonly type: 'Noise';
}

interface Pending {
  readonly type: 'PendingCharacter';
}

interface Letter {
  readonly type: 'Letter';
  readonly value: string;
}

type Char = Noise | Letter | Pending;

export type Type = Char;

/**
 * Default constructor for noise
 */
export const noise = { type: 'Noise' } as Noise;
/**
 * Default constructor for pending char
 */
export const pendingChar = { type: 'PendingCharacter' } as Pending;

/**
 * Construct a letter type with a given string as its content
 * @param l a string, expects a single letter or syllabic unit
 */
export const constructLetter = (l: string) => {
  if (l !== '#noise') {
    return {
      type: 'Letter',
      value: l,
    } as Letter;
  } else {
    return noise;
  }
};

/**
 * Says if a char is a letter
 * @param c the char
 */
export const isLetter = (c: Char) => {
  return c.type === 'Letter';
};

/**
 * Says if a char is pending char
 * @param c the char
 */
export const isPending = (c: Char) => {
  return c.type === 'PendingCharacter';
};

/**
 * Says if a char is a noise
 * @param c the char
 */
export const isNoise = (c: Char) => {
  return c.type === 'Noise';
};

/**
 * Get the inner content of a char, if it's a noise, return '#noise', if it's pending, return undefined
 * @param x the char
 */
export const getChar = (x: Char): string | undefined => {
  if (x.type === 'Letter') {
    return x.value;
  } else if (x.type === 'Noise') {
    return '#noise';
  } else {
    return undefined;
  }
};
