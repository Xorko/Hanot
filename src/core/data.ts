/* Data type definitions for the usage of convert an inkml file (string) into a json object or convert it back
 * `Chan`, `AnnoData`, `TraceGroupData` and `InkData` for internal usage
 * `Data` for usage with xml-parser
 */

export const enum Chan {
  X = 'X',
  Y = 'Y',
  F = 'F',
  T = 'T',
}

export interface AnnoData {
  '#text': string;
  attr: { type: string };
}

export interface TraceGroupData {
  attr?: {
    'xml:id'?: string;
    noise?: string;
    positionInGroundTruthValue?: number;
    [key: string]: any;
  };
  annotationXML?: {
    attr: { type: string };
    annotation: AnnoData[];
  };
  annotation?: AnnoData[];
  traceGroup?: TraceGroupData | TraceGroupData[];
  trace?: string | string[];
}

export interface InkData {
  attr: {
    xmlns: string;
  };
  traceFormat: {
    channel: {
      attr: {
        name: Chan;
        type: 'decimal' | 'integer';
      };
    }[];
  };
  annotation?: AnnoData[];
  traceGroup: TraceGroupData | TraceGroupData[];
}

interface Data {
  ink: InkData;
}

export type Type = Data;
