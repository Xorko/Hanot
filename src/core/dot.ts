interface Dot {
  readonly x: number;
  readonly y: number;
  readonly f: number;
  readonly t: number;
}

export type Type = Dot;
