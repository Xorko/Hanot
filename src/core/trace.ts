import * as Dot from './dot';

interface Trace {
  dots: Dot.Type[];
}

export type Type = Trace;
