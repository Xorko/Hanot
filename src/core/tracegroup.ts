import * as Trace from './trace';
import { Char } from './char';

interface TraceGroup {
  readonly traces: Trace.Type[];
  label: Char;
}

export type Type = TraceGroup;
