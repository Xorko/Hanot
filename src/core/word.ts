import type { SerializableMap } from '../screens/file-selection-screen/types/file-import-types';
import * as Trace from './trace';
import * as TraceGroup from './tracegroup';
import * as XMLAnnotation from './xml-annotation';

interface Word {
  tracegroups: TraceGroup.Type[];
  readonly annotationsXML?: XMLAnnotation.Type;
  readonly annotations: SerializableMap<string>;
  readonly attributes: SerializableMap<any>;

  readonly predicted?: string;

  defaultTraceGroup: Trace.Type[];
}

export type Type = Word;
