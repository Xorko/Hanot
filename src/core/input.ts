import isEmpty from 'lodash/isEmpty';
import { SerializableMap } from '../types/core-types';
import * as Char from './char';
import * as Data from './data';
import { TraceData } from './data';
import * as Dot from './dot';
import * as InkML from './inkml';
import * as Trace from './trace';
import * as TraceGroup from './tracegroup';
import { createEmptyTraceGroup } from './tracegroup';
import * as Word from './word';

/**
 * Construct an InkML type from the raw json data converted from an inkml file.
 * @param ink the InkData json object to be converted, do nothing if it's undefined.
 */
export const constructData = (ink?: Data.InkData): InkML.Type | undefined => {
  if (ink !== undefined) {
    const anno = makeAnnotations(ink.annotation);
    return {
      words: makeWords(ink.traceGroup, anno),
    };
  } else {
    return undefined;
  }
};

const makeWords = (
  tg: Data.TraceGroupData | Data.TraceGroupData[],
  anno?: SerializableMap<string>,
): Word.Type[] => {
  if (Array.isArray(tg)) {
    return tg.map(e => makeSingleWord(e, anno));
  } else {
    return [makeSingleWord(tg, anno)];
  }
};

const makeSingleWord = (
  tg: Data.TraceGroupData,
  anno?: SerializableMap<string>,
): Word.Type => {
  const predicted = tg.attr?.['xml:id'];
  const attrs = {} as SerializableMap<any>;
  if (tg.attr !== undefined) {
    for (const [k, v] of Object.entries(tg.attr)) {
      if (!['xml:id', 'positionInGroundTruthValue', 'noise'].includes(k)) {
        attrs[k] = v;
      }
    }
  }
  let annoXML;
  if (tg.annotationXML !== undefined) {
    annoXML = {
      type: tg.annotationXML.attr.type,
      values:
        makeAnnotations(tg.annotationXML.annotation) ??
        ({} as SerializableMap<string>),
    };
  }
  const [traceGroups, danglingTraces] = constructTraceGroups(tg);
  return {
    tracegroups: traceGroups,
    annotationsXML: annoXML,
    annotations:
      anno ?? makeAnnotations(tg.annotation) ?? ({} as SerializableMap<string>),
    attributes: attrs,
    predicted: predicted,
    defaultTraceGroup: [...danglingTraces, ...constructDefaultTraceGroup(tg)],
  };
};

const makeAnnotations = (
  annos?: Data.AnnoData[],
): SerializableMap<string> | undefined => {
  const ret: SerializableMap<string> = {};

  for (const anno of annos ?? []) {
    if (anno.attr !== undefined) {
      ret[anno.attr.type] = anno['#text'];
    }
  }

  return isEmpty(ret) ? undefined : ret;
};

const constructTraceGroups = (
  tg: Data.TraceGroupData,
): [TraceGroup.Type[], Trace.Type[]] => {
  if (Array.isArray(tg.traceGroup)) {
    return constructTraceGroupFromTraceGroupArray(tg.traceGroup);
  } else if (tg.traceGroup !== undefined) {
    const ret = constructTraceGroupFromTraceGroup(tg.traceGroup);
    return [ret[0], ret[2]];
  } else {
    return [[], []];
  }
};

const constructDefaultTraceGroup = (tg: Data.TraceGroupData): Trace.Type[] => {
  if (Array.isArray(tg.trace)) {
    return tg.trace.map(x => constructTrace(x));
  } else if (tg.trace !== undefined) {
    return [constructTrace(tg.trace)];
  } else {
    return [];
  }
};

const constructTraceGroupFromTraceGroupArray = (
  tg: Data.TraceGroupData[],
): [TraceGroup.Type[], Trace.Type[]] => {
  const nullableTraceGroupWithIndex = tg.map(c =>
    constructTraceGroupFromTraceGroup(c),
  );
  const availableTraceGroups = nullableTraceGroupWithIndex.filter(
    e => e[1].length !== 0,
  );
  const danglingTraces = nullableTraceGroupWithIndex
    .filter(e => e[2].length !== 0)
    .flatMap(e => e[2]);
  return [reconstructTracesInOrder(availableTraceGroups), danglingTraces];
};

const constructTraceGroupFromTraceGroup = (
  tg: Data.TraceGroupData,
): [TraceGroup.Type[], number[], Trace.Type[]] => {
  const xmlid = tg.attr?.['xml:id'];
  const pos = tg.attr?.positionInGroundTruthValue;
  const noise = tg.attr?.noise;
  let trace: Trace.Type[] = [];
  if (Array.isArray(tg.trace)) {
    trace = tg.trace.map(constructTrace);
  } else if (tg.trace !== undefined) {
    trace = [constructTrace(tg.trace)];
  }
  if (pos !== undefined && xmlid !== undefined && noise === undefined) {
    return [
      [
        {
          traces: trace,
          label: Char.constructLetter(xmlid),
        },
      ],
      [pos],
      [],
    ];
  } else if (pos !== undefined && xmlid === undefined && noise !== undefined) {
    return [
      [
        {
          traces: trace,
          label: Char.noise,
        },
      ],
      [pos],
      [],
    ];
  } else if (pos === undefined && xmlid === undefined && noise === undefined) {
    return [[], [], trace];
  } else {
    throw new Error('Impossible case');
  }
};

const constructTrace = (tr: string | TraceData): Trace.Type => {
  if (typeof tr === 'string') {
    return {
      dots: constructTraceFromString(tr),
      oldTrace: -1,
    };
  } else {
    return {
      dots: constructTraceFromString(tr['#text']),
      oldTrace: parseInt(tr.attr.oldTrace, 10),
    };
  }
};

const constructTraceFromString = (s: string) => {
  if (!s) {
    return [];
  }

  return s
    .split(',')
    .map(xxs =>
      xxs
        .split(' ')
        .filter(e => e !== '')
        .map(e => parseFloat(e)),
    )
    .map(constructDot);
};

const constructDot = (nx: number[]): Dot.Type => {
  return {
    x: nx[0],
    y: nx[1],
    f: nx[2],
    t: nx[3],
  };
};

const reconstructTracesInOrder = (
  tdata: [TraceGroup.Type[], number[], Trace.Type[]][],
): TraceGroup.Type[] => {
  const nonEmptyTraceGroups = tdata.filter(
    e => e[1] !== undefined && e[2] !== undefined,
  );
  const rightmost = max(nonEmptyTraceGroups.map(e => e[1][0]));
  const traceGroupForms: [TraceGroup.Type, number][] = nonEmptyTraceGroups.map(
    e => [e[0][0], e[1][0]],
  );
  const orderedTraceGroups: TraceGroup.Type[] = [];
  if (rightmost !== undefined) {
    for (const i of Array.from(Array(rightmost + 1).keys())) {
      const curr = traceGroupForms.find(e => e[1] === i);
      if (curr !== undefined) {
        orderedTraceGroups.push(curr[0]);
      } else {
        orderedTraceGroups.push(createEmptyTraceGroup());
      }
    }
  }
  return orderedTraceGroups;
};

const max = (a: number[]): number | undefined => {
  if (a.length === 0) {
    return undefined;
  }
  let maxValue = a[0];
  for (const x of a) {
    if (x > maxValue) {
      maxValue = x;
    }
  }
  return maxValue;
};
