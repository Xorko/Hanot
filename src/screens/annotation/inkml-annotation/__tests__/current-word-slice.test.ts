import { merge } from 'lodash';
import { constructLetter } from '../../../../core/char';
import * as Dot from '../../../../core/dot';
import { createEmptyTraceGroup } from '../../../../core/input';
import * as Trace from '../../../../core/trace';
import * as TraceGroup from '../../../../core/tracegroup';
import * as Word from '../../../../core/word';
import currentWordReducer, {
  annotateTraceGroup,
  deleteTraceGroups,
  initWord,
  pushDots,
  pushTraceGroup,
  pushTraces,
  setDefaultTraceGroup,
  setFinalTraceGroups,
} from '../current-word-slice';

const initialWord: Word.Type = {
  tracegroups: [],
  defaultTraceGroup: [],
  annotations: {},
  attributes: {},
  predicted: undefined,
};

const customWord1: Word.Type = {
  tracegroups: [
    {
      traces: [],
      label: constructLetter('a'),
    },
    {
      traces: [],
      label: constructLetter('b'),
    },
  ],
  defaultTraceGroup: [
    {
      dots: [{ x: 5, y: 5, f: 3, t: 2 }],
      oldTrace: 2,
    },
  ],
  annotations: {},
  attributes: {},
  predicted: undefined,
};

function createFactory<T>(data: T) {
  return (...args: Partial<T>[]): T => {
    return merge({}, data, ...args);
  };
}

const factoryInitialWord = createFactory(initialWord);
const factoryCustomWord1 = createFactory(customWord1);

//afterEach(cleanup);

test('Test initWord', () => {
  const expectedWord = factoryInitialWord();

  expect(currentWordReducer(undefined, initWord(initialWord))).toEqual(
    expectedWord,
  );
});

test('Test pushTraces with empty array and a currentWord with no traceGroup in traceGroups', () => {
  const traces: Trace.Type[] = [];

  expect(() =>
    currentWordReducer(initialWord, pushTraces(traces)),
  ).toThrowErrorMatchingSnapshot();
});

test('Test pushTraces with one element array and a currentWord with no traceGroup in traceGroups', () => {
  const traces: Trace.Type[] = [
    {
      dots: [],
      oldTrace: 0,
    },
  ];

  expect(() =>
    currentWordReducer(initialWord, pushTraces(traces)),
  ).toThrowErrorMatchingSnapshot();
});

test('Test pushTraces with one element array and a currentWord with a traceGroup in traceGroups', () => {
  const traces: Trace.Type[] = [
    {
      dots: [],
      oldTrace: 0,
    },
  ];
  const expectedWord = factoryCustomWord1({
    tracegroups: [
      customWord1.tracegroups[0],
      {
        traces: traces,
        label: customWord1.tracegroups[1].label,
      },
    ],
  });
  expect(currentWordReducer(customWord1, pushTraces(traces))).toEqual(
    expectedWord,
  );
});

test('Test setDefaultTraceGroup', () => {
  const traces: Trace.Type[] = [
    {
      dots: [],
      oldTrace: 0,
    },
  ];
  const expectedWord = factoryInitialWord({
    defaultTraceGroup: traces,
  });
  expect(currentWordReducer(initialWord, setDefaultTraceGroup(traces))).toEqual(
    expectedWord,
  );
});

test('Test setDefaultTraceGroup on non-empty defaultTraceGroup field', () => {
  const traces: Trace.Type[] = [
    {
      dots: [],
      oldTrace: 0,
    },
  ];
  const expectedWord = factoryCustomWord1();
  expectedWord.defaultTraceGroup = traces;
  expect(currentWordReducer(customWord1, setDefaultTraceGroup(traces))).toEqual(
    expectedWord,
  );
});

test('Test setFinalTraceGroups', () => {
  const finaltraceGroups: TraceGroup.Type[] = [
    {
      traces: [],
      label: constructLetter('c'),
    },
  ];
  const expectedWord = factoryInitialWord({
    tracegroups: finaltraceGroups,
  });
  expect(
    currentWordReducer(initialWord, setFinalTraceGroups(finaltraceGroups)),
  ).toEqual(expectedWord);
});

test('Test setFinalTraceGroups on non-empty finalTraceGroups field', () => {
  const finaltraceGroups: TraceGroup.Type[] = [
    {
      traces: [],
      label: constructLetter('c'),
    },
  ];
  const expectedWord = factoryCustomWord1();
  expectedWord.tracegroups = finaltraceGroups;
  expect(
    currentWordReducer(customWord1, setFinalTraceGroups(finaltraceGroups)),
  ).toEqual(expectedWord);
});

test('Test deleteTraceGroups if oldTrace is out of bound of defaultTraceGroup', () => {
  const traceGroup: TraceGroup.Type = {
    traces: [
      {
        dots: [{ x: 5, y: 5, f: 3, t: 2 }],
        oldTrace: 2,
      },
    ],
    label: constructLetter('a'),
  };
  const expectedWord = factoryCustomWord1();
  expectedWord.defaultTraceGroup[traceGroup.traces[0].oldTrace] = {
    dots: traceGroup.traces[0].dots,
    oldTrace: -1,
  };

  expect(
    currentWordReducer(customWord1, deleteTraceGroups([traceGroup])),
  ).toEqual(expectedWord);
});

test('Test deleteTraceGroups', () => {
  const traceGroup: TraceGroup.Type = {
    traces: [
      {
        dots: [{ x: 6, y: 5, f: 3, t: 2 }],
        oldTrace: 0,
      },
    ],
    label: constructLetter('a'),
  };
  const expectedWord = factoryCustomWord1();
  expectedWord.defaultTraceGroup = [
    {
      dots: [
        { x: 6, y: 5, f: 3, t: 2 },
        { x: 5, y: 5, f: 3, t: 2 },
      ],
      oldTrace: 2,
    },
  ];
  expect(
    currentWordReducer(customWord1, deleteTraceGroups([traceGroup])),
  ).toEqual(expectedWord);
});

test('Test pushTraceGroup', () => {
  const expectedWord = factoryCustomWord1();
  expectedWord.tracegroups.push(createEmptyTraceGroup());
  expect(currentWordReducer(customWord1, pushTraceGroup())).toEqual(
    expectedWord,
  );
});

test('Test pushDots with incorrect idxTraceGroup parameter', () => {
  const leftTrace: Dot.Type[] = [{ x: 5, y: 5, f: 3, t: 2 }];
  const idxOldTrace: number = 6;
  const idxTraceGroup: number = -1;
  expect(() =>
    currentWordReducer(
      customWord1,
      pushDots({ leftTrace, idxOldTrace, idxTraceGroup }),
    ),
  ).toThrowErrorMatchingSnapshot();
});

test('Test pushDots', () => {
  const leftTrace: Dot.Type[] = [{ x: 5, y: 5, f: 3, t: 2 }];
  const idxOldTrace: number = 6;
  const idxTraceGroup: number = 1;
  const expectedWord = factoryCustomWord1();
  expectedWord.tracegroups[1].traces[0] = {
    dots: leftTrace,
    oldTrace: idxOldTrace,
  };
  expect(
    currentWordReducer(
      customWord1,
      pushDots({ leftTrace, idxOldTrace, idxTraceGroup }),
    ),
  ).toEqual(expectedWord);
});

test('Test annotateTraceGroup with incorrect index parameter', () => {
  const index: number = -1;
  const annotation: string = 'a';

  expect(() =>
    currentWordReducer(customWord1, annotateTraceGroup({ index, annotation })),
  ).toThrowErrorMatchingSnapshot();
});

test('Test annotateTraceGroup', () => {
  const index: number = 1;
  const annotation: string = 'a';
  const expectedWord = factoryCustomWord1();
  expectedWord.tracegroups[1].label = constructLetter(annotation);
  expect(
    currentWordReducer(customWord1, annotateTraceGroup({ index, annotation })),
  ).toEqual(expectedWord);
});
