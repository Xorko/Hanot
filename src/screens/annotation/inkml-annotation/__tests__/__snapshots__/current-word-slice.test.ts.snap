// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test annotateTraceGroup with incorrect index parameter 1`] = `"Index out of range"`;

exports[`Test pushDots with incorrect idxTraceGroup parameter 1`] = `"Index out of range (idxTraceGroup)"`;

exports[`Test pushTraces with empty array and a currentWord with no traceGroup in traceGroups 1`] = `"Index out of range (state.tracegroups.length - 1]"`;

exports[`Test pushTraces with one element array and a currentWord with no traceGroup in traceGroups 1`] = `"Index out of range (state.tracegroups.length - 1]"`;
